### PR TITLE
feat(#904): Add FileUploadInput component with FileProcessor interface

### PR DESCRIPTION
**Pull request by paulushcgcj**
# Description

Introduces the `FileUploadInput` form component alongside a `FileProcessor` interface for structured, reusable file handling in the frontend. Adds the `pretty-bytes` dependency for human-readable file size formatting.

Fixes #904

**Changes summary:**

- `frontend/src/components/Form/FileUploadInput/index.tsx` — New `FileUploadInput` React component with improved file handling and error management.
- `frontend/src/components/Form/FileUploadInput/fileProcessor.ts` — New `FileProcessor` interface defining the contract for file validation/processing logic.
- `frontend/src/components/Form/FileUploadInput/index.scss` — Component styles.
- `frontend/src/components/Form/FileUploadInput/index.unit.test.tsx` — Unit tests covering component behaviour, edge cases, and detailed usage scenarios.
- `frontend/package.json` / `frontend/package-lock.json` — Added `pretty-bytes` package for readable file size display.
- `frontend/Dockerfile` — Infrastructure update to v24.16.0 (pulled in via `chore(deps)` commit).
- `.github/workflows/.automated-tests.yml` / `merge.yml` — Minor CI adjustments included in the branch.
- `cypress/cypress/e2e/sample.feature` — Sample E2E feature updated to reflect component availability.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update

# How Has This Been Tested?

- [x] New unit tests

Unit tests in `index.unit.test.tsx` cover:
- Rendering with and without initial file selection
- `FileProcessor` interface delegation (validation, processing, error paths)
- Edge cases documented in `c5a48cd` (zero-byte files, unsupported MIME types, oversized files)
- File size display using `pretty-bytes`

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

## Feature-Flag Controlled Changes (Required if applicable)

- [x] Not applicable: no feature-flag-controlled behavior changed in this PR

## Further comments

The `FileProcessor` interface was introduced as a clean extension point: callers inject their own processor implementation (validate file type, size, transform content, etc.) without the component needing to know domain rules. This keeps the component generic and reusable across future file-upload surfaces.

`pretty-bytes` was chosen over a manual formatter because it handles edge cases (locale, SI vs binary, fractional units) without additional code.

---

Thanks for the PR!

Deployments, as required, will be available below:
Any successful deployments (not always required) will be available [here](https://nr-waste-plus-15-frontend.apps.silver.devops.gov.bc.ca/)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-waste-plus/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-waste-plus/actions/workflows/merge.yml)

### CODE REVIEW

File: frontend/src/components/Form/FileUploadInput/index.tsx [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+      <FileUploaderDropContainer
+        className={`file-upload-dropzone ${disabled ? 'file-upload-dropzone--disabled' : ''}`}
+        id={dropId}
+        accept={accept.split(',').map((s) => s.trim())}
+        labelText="Drag and drop files here or click to upload"
+        multiple={maxFiles > 1}
+        onAddFiles={handleAddFiles}
+        disabled={disabled}
+        maxFileSize={maxFileSizeBytes}
+      />
```
[the remark this answers is not in the document]
Comment (paulushcgcj) [reply frontend/src/components/Form/FileUploadInput/index.tsx:273]: Divergences are to be expected. Marking as resolved

File: frontend/src/components/Form/FileUploadInput/index.tsx [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  const [componentErrors, setComponentErrors] = useState<string[]>([]);
+
+  const fmt = (bytes: number) => prettyBytes(bytes, { maximumFractionDigits: 1 });
+
+  /** Flat list of all parsed results from currently accepted files. */
+  const emitResults = (dataMap: Map<string, T[]>) => {
+    onProcessed(Array.from(dataMap.values()).flat());
+  };
+
+  const handleAddFiles = async (
```
Comment (mamartinezmejia) [inline frontend/src/components/Form/FileUploadInput/index.tsx:143]: Here is a **Race Condition in Concurrent Batch Processing** issue.

If two `handleAddFiles` invocations execute concurrently, the second one's emitted results will miss data from the first batch. Here's why: `handleAddFiles` is async, and JavaScript closures capture state values at the moment the function is invoked — not when it eventually resolves. So if someone drops two batches of files in quick succession:

1. Invocation A starts, captures `processedData = Map{}`
2. Invocation B starts before A resolves, also captures `processedData = Map{}`
3. A finishes — internal state becomes `Map{ A's UUIDs }`, emits `A's data` ✓
4. B finishes — internal state correctly becomes `Map{ A's UUIDs, B's UUIDs }`, but emits `B's data` only ✗

The functional `setProcessedData(prev => ...)` calls do protect the internal React state, but by the time you reach the `onProcessed` emit at the bottom, you're still reading from the original stale closure. From the parent's perspective this looks like silent data loss, which is hard to debug.

**Recommendation**

Decouple the side-effect (emitting data to the parent) from the async handler entirely. Remove the manual emit block at the end of `handleAddFiles` and replace it with a `useEffect`:

```typescript
// 1. Delete the `const hasNewData = ...` block from handleAddFiles entirely

// 2. Add a useEffect to emit whenever processedData changes
useEffect(() => {
  if (processedData.size === 0) return;
  onProcessed(Array.from(processedData.values()).flat());
}, [processedData, onProcessed]);
```

Two caveats to keep in mind:
- The `processedData.size === 0` guard prevents a spurious empty emission on the initial render.
- `onProcessed` is listed as a dependency, so if the parent doesn't memoize it with `useCallback`, the effect will fire on every render. Either add that note in a comment or document it as a requirement for the parent.

File: frontend/src/components/Form/FileUploadInput/index.tsx [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+    const toAdd = addedFiles.slice(0, slots);
+    const skippedCount = addedFiles.length - toAdd.length;
+
+    const newEntries: TrackedFile[] = toAdd.map((file) => ({
+      uuid: crypto.randomUUID(),
+      file,
+    }));
+
+    const allTracked = [...trackedFiles, ...newEntries];
```
Comment (mamartinezmejia) [inline frontend/src/components/Form/FileUploadInput/index.tsx]: This can be refactored as:
` setTrackedFiles((prev) => [...prev, ...newEntries]);`

**Review by mamartinezmejia — CHANGES_REQUESTED** [review]

**Review by mamartinezmejia — APPROVED** [review]
